src/object/obj_tx.c

```diff
@@ -2303,20 +2309,23 @@ dc_tx_add_punch_obj(struct dc_tx *tx, daos_handle_t oh, uint64_t flags)
 }
 
 static int
-dc_tx_add_punch_dkey(struct dc_tx *tx, daos_handle_t oh, uint64_t flags,
+dc_tx_add_punch_dkey(struct dc_tx *tx, struct dc_object **obj, uint64_t flags,
 		     daos_key_t *dkey)
 {
 	struct daos_cpd_sub_req	*dcsr;
-	struct dc_object	*obj = NULL;
 	int			 rc;
 
+	D_ASSERT(obj != NULL);
+
+	if (*obj == NULL)
+		return -DER_NO_HDL;
+
 	rc = dc_tx_get_next_slot(tx, false, &dcsr);
 	if (rc != 0)
 		return rc;
 
-	obj = dcsr->dcsr_obj = obj_hdl2ptr(oh);
-	if (dcsr->dcsr_obj == NULL)
-		return -DER_NO_HDL;
+	dcsr->dcsr_obj = *obj;
+	*obj = NULL;
```

Comment:
not sure if can directly assign it as NULL, for example need not dec_ref(obj)?

Reply:
"dec_ref(obj)?" will be done via dec_ref(dcsr->dcsr_obj) when cleanup the TX.